Provide cadvisor stats as prometheus metrics

discordianfish opened the pull request:
This is ~~WIP~~ ready to get merged ~~and provided for discussion:~~

- ~~name is aliases joined by ',' and definitely not what we want: Ideas to solve that? We can~~
 - We're using Aliases[0], if exists, as name
- ~~It doesn't support image name as metric dimension since I'm not sure how to get this~~
 - Out of scope of this PR but should be added eventually
- ~~Unsure about package structure: I didn't use prometheus/ but exporter/ because a) it might not be the only exporter? b) the import path in main would clash with client_golang/prometheus~~
  - we decided for metrics/
- ~~No tests; Will look into that once we agreed on the general form~~
  - Tests will provided once https://github.com/prometheus/client_golang/issues/58 is fixed
- ~~Usage is in seconds, not percent. Not sure what's better~~ Using seconds is fine

Implements #502

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+	networkTxErrors  *prometheus.CounterVec
+	networkTxDropped *prometheus.CounterVec
+
+	tasksSleeping        *prometheus.GaugeVec
```
This should all be one metric I think, as it sounds like a partition on the tasks

discordianfish replied:
It's always hard to answer those things.. Any other opinions? But I'm fine with using one metric.

vmarmol replied:
"tasks_count" broken down by type sounds good.

Although I agree it is a slippery slope. network_bytes broken down by tx, rx. Memory broken down by usage, working set, etc. Those can stay as is :)

brian-brazil replied:
tasks_state sounds like a good name.

http://prometheus.io/docs/practices/naming/#metric-names has guidelines on when to use labels.

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+		lastSeen: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "last_seen",
+			Help:      "Last time a container was seen by the exporter",
```
This sounds like a gauge

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+		),
+		cpuUsageSecondsPerCPU: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "cpu_usage_per_cpu_seconds_total",
```
I'd only have one cpu metric, and split it out by cpu.

You shouldn't export multiple versions of a metric with different labels unless you're running into performance problems.

discordianfish replied:
Good point, will do that.

vmarmol replied:
These are also in nanoseconds

brian-brazil replied:
> These are also in nanoseconds

They're converted down in line 385

vmarmol replied:
Yeah I removed my other comment and not this one...ignore :)

discordianfish replied:
Just realized that CpuStats's PerCpu doesn't include system vs user usage. Only the overall cpu usage is split into user / system.

- Should we keep both metrics?
- Ignore the overall usage and live without system usage?
- Change CpuStats to include system / user usage in PerCpu (obviously the hardest option to implement)

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+		memoryFailures: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "memory_failures_total",
+			Help:      "Number of times memory usage hits limits.",
```
s/hits/hit/

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+		networkTxBytes: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "network_transmit_bytes_total",
+			Help:      "Cumulative count of bytes transmitd",
```
transmitted, and below

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+	defer c.mutex.Unlock()
+
+	for _, v := range c.vectors {
+		v.Reset()
```
I think you'd be better off creating the metrics afresh each time. It avoids locking too.

discordianfish replied:
You mean instead of creating them via NewXX in NewPrometheusCollector() I would create them in here? But I assume I need to create them before Describe() gets called?

brian-brazil replied:
Yes.

I'm used to the other clients where there is no Describe() :)

vmarmol replied:
This is probably ignorance on my part :) but what does Reset() do here?

brian-brazil replied:
Removes all samples from that metric.

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,455 @@
+
+	containers, err := c.manager.SubcontainersInfo("/", &info.ContainerInfoRequest{NumStats: 1})
+	if err != nil {
+		c.errors.Inc()
```
Usually this is done as a gauge that's 1/0 depending on whether this particular scrape worked.

discordianfish replied:
It is? Because this kind of error isn't about processing some items for which a error rate would be useful but something affecting overall scrape? We probably should add that to the best practices in the prometheus docs.

brian-brazil replied:
http://prometheus.io/docs/practices/instrumentation/#collectors :)

discordianfish replied:
yay :)

brian-brazil commented:
> name is aliases joined by ',' and definitely not what we want: Ideas to solve that? We can:

What I'd do is have an additional metric called 'names', with a label called 'name' and a timeseries for each name with value 1.

> Usage is in seconds, not percent. Not sure what's better

Seconds is what you want as you should leave the math to prometheus

discordianfish commented:
@brian-brazil I don't think we want such 'map metric' in this case. The docker container name in aliases is what you are usually use when refering to a container. Always through through that map sounds painful.

beorn7 commented:
General remark about usage of client_golang: If you have metric vec's that you are resetting and filling on each scrape, it should also work to just keep the Desc's around (and return them in Describe()) and then use NewConstMetric to create a throw-away metric to send to the channel in Collect().

Perhaps it's also a good strategy to leave it conceptionally as it is now, and I'll try the approach described above in a separate PR (once we know that the metrics collected are sane).

brian-brazil commented:
> I don't think we want such 'map metric' in this case. The docker container name in aliases is what you are usually use when refering to a container. Always through through that map sounds painful.

It sounds like the names/aliases aren't static, so it'd not wise to use them as label values on everything as every change to them will break your graphs. Having all the aliases may be usable for particular use cases such as Docker but not work for the general case.

This incidentally is where the more advanced label features and console templates come into play.

discordianfish commented:
@vmarmol Do you have some input before I work on the remaining bits and look into getting a CLA signed?

vmarmol commented:
@discordianfish sorry, it's been on my list to take a look and I've been bad at keeping myself to that. I will take a look today :)

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		manager: manager,
+		errors: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "errors_total",
```
nit: maybe metric_errors_total? Not to confuse with some other kind of error.

discordianfish replied:
Keep in mind it's still prefixed by container, so would be 'container_errors_total'. Still, probably not the best name but maybe something like 'container_exporter_errors_total' is better than metric_errors? Other opinions? @brian-brazil usually has a good feeling for those things :)

discordianfish replied:
I've changed it to (container_)export_errors (removed the _total) because it's not a Gauge as suggestd by @brian-brazil

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+			Name:      "cpu_usage_seconds_total",
+			Help:      "Total seconds of cpu time consumed.",
+		},
+			[]string{"name", "id", "image", "type"},
```
It's more useful to expose CPU usage total not broken down since it is significantly more accurate than broken down. We can either have type=total,user,system or have a separate total and breakdown metric. I'm guessing the first is better? If we go that route we can rename to cpu_usage_seconds

brian-brazil replied:
Why do you expect them to be less accurate when broken down? If is this is the same sort of cpu stat as in /proc/stat I'd expect it to all add up.

vmarmol replied:
These are cgroup stats. We get total and per-cpu usage with nanosecond precision. System and user breakdown has jiffy precision which is typically ~10ms.

brian-brazil replied:
Sounds like two different metrics then.

Naming will be fun on those as you want to avoid "by" in metric names as a label will not always be present after aggregation.

discordianfish replied:
Ah, then ignore my comment above and let's just keep both metrics for now? Seem the best we can do. @brian-brazil Naming, do you have some suggestion?

discordianfish replied:
I've added it back now. Let me know if you want me to change the name.

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		fsUsage: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "fs_usage_bytes",
+			Help:      "Number of bytes that is consumed by the container on this filesystem.",
```
nit: s/is/are/

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		fsReadsSectors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "fs_sector_reads_total",
+			Help:      "Number of sectors reads completed",
```
nit: s/sectors/sector/

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		fsWritesSectors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "fs_sector_writes_total",
+			Help:      "Number of sectors writes completed",
```
nit: s/sectors/sector/

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		networkRxErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "network_receive_errors_total",
+			Help:      "Cumulative count of errors encountered",
```
"...while receiving"

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		networkTxDropped: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "network_transmit_packets_dropped_total",
+			Help:      "Cumulative count of bytes transmitd",
```
"Cumulative count of packets dropped"

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		),
+		networkTxErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "network_transmit_errors_total",
```
"...while transmitting"

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+		id := container.Name
+		name := strings.Join(container.Aliases, ",")
+		stats := container.Stats[0]
+		image := "undefined" // FIXME:
```
I think its fine to not use image. We never really expose it in cAdvisor.

discordianfish replied:
Hrmm I think having the image as a dimension is actually very useful. Usually people have their version numbers / git-shas in the image tag, so you often want to group by that to compare one version of your containers to another version.  Do you think it's hard to get the image name with the current design of cadvisor? If so, we can probably leave the image for now, but I would like to see that in the long term.

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,454 @@
+				metric.WithLabelValues(name, id, image, stat.Device).Set(float64(value))
+			}
+		}
+	}
```
Nice compact collector :)

vmarmol commented:
Thanks for the PR @discordianfish! This looks awesome. +1 to keeping in seconds. As per the name/alias and labels, today we only ever have two so we can split by "id", "name", and "alias" unless there is a better solution.

brian-brazil commented:
> today we only ever have two so we can split by "id", "name", and "alias" unless there is a better solution.

Is that the case for every potential user of this code?

Prometheus as of today has the features to work with a single timeseries that has all the potentially useful labels.

vmarmol commented:
@brian-brazil for Docker containers yes, for non-Docker containers no (we'll only have an ID for those). I'm not sure I understand your second sentence?

brian-brazil commented:
> I'm not sure I understand your second sentence?

I'd propose to have metrics such as:
```
container_names{id="myid",name="name1"} 1
container_names{id="myid",name="name2"} 1
container_names{id="myid",name="name3"} 1
other_metrics{id="myid"} 7
```

Using the `group_left`/`group_right` modifiers you can join the names on to the metrics when you need the names, without the disadvantages of dealing with names changing over time.

vmarmol commented:
@brian-brazil we can try that. It does sound slightly painful to write queries for though :) I don't think the names really change over time. Theoretically the alias can, although I haven't seen it in practice.

brian-brazil commented:
> I don't think the names really change over time. Theoretically the alias can, although I haven't seen it in practice.

Okay, sounds like we can get away with the name as a label on everything and an alias as I suggest.

discordianfish commented:
>>  I don't think the names really change over time. Theoretically the alias can, although I haven't seen it in practice.
> Okay, sounds like we can get away with the name as a label on everything and an alias as I suggest.

@brian-brazil What's exactly wrong with changing label values? We still have the id in here which exists as long as the container exists, aka this gets stale after the container is gone. But I assume that's okay, isn't it?:

discordianfish commented:
Okay, I've changed the PR according to the discussion. I also changed it as @beorn7 suggested and did some moving around.
@vmarmol Are you okay with the package structure as it is right now?

brian-brazil commented:
> But I assume that's okay, isn't it?:

It's not okay. You want your timeseries to have the same labels over their lifetime, you'll run into all sorts of problems if this is not the case.

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,350 @@
+		manager: manager,
+		errors: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "export_errors",
```
We call this `scrape_error` elsewhere

discordianfish replied:
But it's not really an error while scraping? But I'm fine with changing it.

discordianfish replied:
I've changed it now to scrape_error with same description as in node-exporter/textfile

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,350 @@
+func (c *PrometheusCollector) Collect(ch chan<- prometheus.Metric) {
+	containers, err := c.manager.SubcontainersInfo("/", &info.ContainerInfoRequest{NumStats: 1})
+	if err != nil {
+		c.errors.Inc()
```
.Set(1) would be clearer

discordianfish commented:
@brian-brazil Hrm.. maybe this is a misunderstanding. The timeseries, for example 'memory usage of container X' will have the same labels for the lifetime of 'container X'. But if you, for example, delete and recreate the container with the same name etc the id and therfor the label changes. I assume this is okay because it's in fact a new container.

brian-brazil commented:
> The timeseries, for example 'memory usage of container X' will have the same labels for the lifetime of 'container X'. But if you, for example, delete and recreate the container with the same name etc the id and therfor the label changes. I assume this is okay because it's in fact a new container.

That's fine, and what you want.

The issue arises if the labels change during the lifetime of container X, such as if the name or aliases change.

brian-brazil commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,350 @@
+			[]string{"name", "id"},
+			nil),
+		cpuUsageSeconds: prometheus.NewDesc(
+			"container_cpu_usage_seconds_total",
```
You're missing the type label here, and the names of metrics need to be unique.

I don't suppose we can just ignore the user vs. system metric?

discordianfish replied:
Good catch, I've fixed it. I think it's important to differentiate user vs system so would rather keep it. On the long run we could try to get the user vs system dimension into the per_cpu metric but not sure that's even exposed by the kernel. For now, I think it's fine to keep both.

brian-brazil replied:
Okay, I never came across a good way to name things like this.

container_cpu_usageclass_seconds_total maybe?

vmarmol replied:
cpu_usage_type_seconds there are 3 labels: total, user, system no?

brian-brazil replied:
> cpu_usage_type_seconds there are 3 labels: total, user, system no?

The label would be 'type' the value would be 'user' or 'system'

You shouldn't mix total in with breakdowns, as that breaks aggregation.

juliusv commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,350 @@
+		tasks: prometheus.NewDesc(
+			"container_tasks_state",
+			"Number of tasks in given state",
+			[]string{"type", "name", "id"},
```
Does `type` indicate the state? Should it be named `state` then?

discordianfish replied:
Oh sure, thanks!

beorn7 commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,350 @@
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const namespace = "container"
```
You are using this in only one place now. So you could get rid of it. (Or include it in the Desc generation...)

beorn7 commented:
NewConstMetric usage looks good to me.

vmarmol commented on exporter/prometheus.go:
```diff
@@ -0,0 +1,348 @@
+			Help:      "1 if there was an error while getting container metrics, 0 otherwise",
+		}),
+		lastSeen: prometheus.NewDesc(
+			"container_last_seen",
```
Don't we not need "container" in the name of all of these due to the namespace?

beorn7 replied:
The namespace is only one of the options in `...Opts`, which are used to construct fully fledged metric objects for Prometheus. Those are meant to be used for "native" Prometheus metrics. What we are doing here is copying existing metrics into Prometheus metrics. For that, we create some of the Prometheus metrics on the fly during scraping, which does not follow the 'high level' path with the Opts object, but assembles the necessary parts one level further down (in this case a metric descriptor is created, which doesn't feature setting a namespace via Opts - so the "container" part has to be included in the name).

vmarmol commented:
GitHub lost the other comment threads so I'll reply to some here (let me know if I missed anything).

About the image. It is awkward to get them with the API today. We can put this in as is today and see how to get the image data in.

For the package structure, can we rename "exporter" to "metrics". We will probably use it for more metrics.

juliusv commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,348 @@
+			nil),
+		fsWeightedIoTime: prometheus.NewDesc(
+			"container_fs_io_time_weighted_seconds_total",
+			"Number of seconds spent doing I/Os",
```
Should this and some of the counters above also start with "Cumulative count of...", like the counters below? That makes it clear to the user that they are indeed cumulative and not over some arbitrary period of time.

juliusv commented:
:+1: It's potentially possible to simplify the metric instantiation/mapping code in the future, but looks good for now.

vmarmol commented:
This looks good @discordianfish! Can we rebase, and I think we can merge :)

vmarmol commented:
LGTM, will merge on green. A million thanks @discordianfish :)

brian-brazil commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,348 @@
+			[]string{"name", "id", "type"},
+			nil),
+		cpuUsageSecondsPerCPU: prometheus.NewDesc(
+			"container_cpu_usage_seconds_total",
```
This needs to be named differently to the above metric, otherwise prometheus won't be able to parse the metrics

vmarmol replied:
"container_per_cpu_usage_seconds_total"?

brian-brazil replied:
That'll be confusing when aggregated up, i'd suggest container_cpu_usagetype_seconds_total for the other

discordianfish replied:
Good catch, seems like I missed that when rebasing. I'm not very happy with container_cpu_(usagetype)_seconds_total but it's probably better the best option.

vmarmol commented:
ok to test

vmarmol commented:
Merged! Thanks @discordianfish :D
